convert reactor-config-map plugin into core config module

rcerven opened the pull request:
* CLOUDBLD-6304

Signed-off-by: Robert Cerven <rcerven@redhat.com>

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [x] JSON/YAML configuration changes are updated in the relevant schema
- [x] Changes to metadata also update the documentation for the metadata
- [ ] Pull request has a link to an osbs-docs PR for user documentation updates
- [x] New feature can be disabled from a configuration file

MartinBasti commented on atomic_reactor/config.py:
```diff
@@ -0,0 +1,567 @@
+        cluster_configs = self._get_cluster_configuration()
+        return platform in cluster_configs
+
+    def get_odcs_config(self):
```
Property?

rcerven replied:
sure

MartinBasti commented on atomic_reactor/config.py:
```diff
@@ -0,0 +1,567 @@
+            all_cluster_configs[platform] = cluster_configs
+        return all_cluster_configs
+
+    def get_enabled_clusters_for_platform(self, platform):
```
Property?

rcerven replied:
it needs platform arg

MartinBasti commented on atomic_reactor/config.py:
```diff
@@ -0,0 +1,567 @@
+
+    @property
+    def koji_path_info(self):
+        from koji import PathInfo
```
Why import is here?

rcerven replied:
it was like that in reactor_config_map plugin, but I can move all those imports to the top if you want

MartinBasti commented on atomic_reactor/build.py:
```diff
@@ -9,16 +9,11 @@
 Classes which implement tasks which builder has to be capable of doing.
 Logic above these classes has to set the workflow itself.
 """
-import re
```
Why changes in build.py were done?

rcerven replied:
in config I need dockerfileimages, which were in InsideBuilder (which will be removed), so I've already moved useful attributes to workflow

        self.df_dir = build_file_dir
        self._df_path = None
        self.original_df = None
        self.buildargs = {}  # --buildargs for container build
        self.dockerfile_images = DockerfileImages([])

MartinBasti replied:
Why dockerfile images must be in config?

rcerven replied:
they are only used there to set source registry & organization 
https://github.com/containerbuildsystem/atomic-reactor/pull/1654/files#diff-3079e0bf8bc543b512b09e347622e9a2457237692c644efe6319d113c0ef4f1eR185

rcerven commented:
input & build plugins (except for orchestrate & source) removed

also moved from InsideBuilder to workflow

    self.df_dir = build_file_dir
    self._df_path = None
    self.original_df = None
    self.buildargs = {}  # --buildargs for container build
    self.dockerfile_images = DockerfileImages([])

as InsideBuilder will be removed

schema now requires also : "openshift", "registries", "source_registry"
as those are mandatory for osbs anyway

removed fallbacks from plugins as well, as all configuration should be read from reactor_config_map

chmeliik commented on atomic_reactor/config.py:
```diff
@@ -0,0 +1,568 @@
+    DEFAULT_CONFIG = {ReactorConfigKeys.VERSION_KEY: 1}
+
+    def __init__(self, workflow, config_path=None, basename=REACTOR_CONFIG_BASE_NAME,
+                 env_name=REACTOR_CONFIG_ENV_NAME, raw_config=None):
```
Do we need all these ways to init a Config? From a quick look, it seems that only `config_path` and `raw_config` are ever used, and `raw_config` only for unit tests.

How about something like this
```python
class Configuration:
    def __init__(self, workflow, raw_config: dict):
        ...

    @classmethod
    def from_file(cls, workflow, path: str) -> Configuration:
        ...
```
(also, I'm not convinced that `workflow` should be there, see next comment)

rcerven replied:
I did leave reading from env variable be for now

raw_config is just for testing purposes

I prefer to have it as params and call all inits the same way

chmeliik commented on atomic_reactor/config.py:
```diff
@@ -0,0 +1,568 @@
+        if self.workflow.dockerfile_images:
+            source_registry_docker_uri = self.source_registry['uri'].docker_uri
+            organization = self.registries_organization
+            self.workflow.dockerfile_images.set_source_registry(source_registry_docker_uri,
```
Does `workflow` need to be an attribute of Configuration? That seems backwards to me (configuration doesn't have a workflow, the workflow has some configuration). How about instead giving this method a `workflow` param? The `openshift_session` property would also need to be changed, probably turned into a method with a `namespace` param.

rcerven replied:
workflow has configuration, which uses/changes workflow

also it is easy to use it like that

chmeliik replied:
Easy to use but IMO harder to understand and keep clean. But I'd like some more opinions on this

MartinBasti replied:
I agree with Adam here.

rcerven replied:
workflow removed

chmeliik commented on atomic_reactor/plugins/post_push_operator_manifest.py:
```diff
@@ -1,172 +0,0 @@
-)
-
-
-class PushOperatorManifestsPlugin(PostBuildPlugin):
```
Could the OMPS removal have its own commit (or maybe its own PR)? And do we want to remove `utils.omps` as well?

chmeliik replied:
And docs will need to be updated too.

rcerven replied:
yeah I'll remove utils.omps as well

I'll move it to separate commit

rcerven replied:
moved

chmeliik commented on atomic_reactor/config.py:
```diff
@@ -0,0 +1,568 @@
+        root_url = self.koji['root_url'].rstrip('/')
+        return PathInfo(topdir=root_url)
+
+    @property
+    def pnc(self):
+        return self._get_value(ReactorConfigKeys.PNC_KEY)
+
+    @property
+    def odcs(self):
+        return self._get_value(ReactorConfigKeys.ODCS_KEY, fallback={})
```
It will be confusing if some properties have fallbacks and others don't.
```python
>>> conf = Configuration(workflow, raw_config={})
>>> conf.odcs
{}
>>> conf.pnc
KeyError: 'pnc'  # why does one fail and the other does not, if neither was in the config?
```
Could this be done more consistently?

rcerven replied:
that is based on schema default values

before we had fallback randomly used in plugins, now we have it only based on schema default values, and without forcing it

chmeliik replied:
In the schema, these top-level properties have a `default` value:
```
group_manifests
prefer_schema1_digest
skip_koji_check_for_base_image
deep_manifest_list_inspection
fail_on_digest_mismatch
list_rpms_from_scratch
allow_multiple_remote_sources
```
A lot of properties have fallbacks and no default in the schema

rcerven replied:
ok I'll add fallback to everything except of required keys in the schema

rcerven replied:
changed

rcerven commented on atomic_reactor/plugin.py:
```diff
@@ -445,12 +445,13 @@ def __init__(self, dt, workflow, plugin_conf, *args, **kwargs):
-            source_method = workflow.builder.source.config.image_build_method
-            system_method = workflow.default_image_build_method
-            plugin_conf = [{'name': source_method or system_method, 'is_allowed_to_fail': False}]
+# no build plugins
+#        else:
+#            # if no buildstep_plugins configured, which is typical for worker builds,
+#            # use what the source says or the system default.
+#            source_method = workflow.source.config.image_build_method
+#            system_method = workflow.default_image_build_method
+#            plugin_conf = [{'name': source_method or system_method, 'is_allowed_to_fail': False}]
```
I'll remove it

MartinBasti commented on atomic_reactor/config.py:
```diff
@@ -0,0 +1,568 @@
+        return self._get_value(ReactorConfigKeys.CACHITO_KEY, fallback={})
+
+    @property
+    def cachito_session(self):
```
`*_session` methods shouldn't be part of class, you can keep them in module but it's very weird having session property as part of configuration

rcerven replied:
*_session moved outside of class as standalone functions

chmeliik commented on atomic_reactor/schemas/config.json:
```diff
@@ -280,38 +230,6 @@
         "additionalProperties": false,
         "required": ["api_url", "auth", "signing_intents", "default_signing_intent"]
     },
-    "omps": {
```
Should this also go in the Remove OMPS commit?

rcerven replied:
moved there

rcerven commented on atomic_reactor/config.py:
```diff
@@ -0,0 +1,565 @@
+    def __init__(self, workflow, config_path=None, basename=REACTOR_CONFIG_BASE_NAME,
+                 env_name=REACTOR_CONFIG_ENV_NAME, raw_config=None):
+
+        self.reactor_config_env = env_name
+        self.reactor_config_path = None
+        if config_path and basename:
+            self.reactor_config_path = os.path.join(config_path, basename)
```
they don't have to, I thought it might be useful to have it accessible

rcerven replied:
removed from self

rcerven commented on atomic_reactor/config.py:
```diff
@@ -0,0 +1,565 @@
+
+        logger.info("reading config content %s", self.conf)
+
+        self.set_workflow_based_on_config(workflow)
```
well, it should be called right after init, so why not have it in the init already

chmeliik replied:
Because then you need a worklfow object to create a Configuration object. It should be possible create a Configuration independently of workflow.

rcerven replied:
ok I'll remove it from there

chmeliik commented on tests/test_config.py:
```diff
@@ -0,0 +1,1470 @@
+        assert isinstance(operator_config, dict)
+        assert "allowed_registries" in operator_config
+
+    def test_set_user_params(self, tmpdir):
+        filename = os.path.join(str(tmpdir), 'config.yaml')
+        with open(filename, 'w') as fp:
+            fp.write(dedent(REQUIRED_CONFIG))
+
+        workflow = self.prepare()
+        Configuration(config_path=str(tmpdir))
+
+        for k, v in USER_PARAMS.items():
+            assert workflow.user_params[k] == v
```
This doesn't test anything related to configuration, does it?

rcerven replied:
right moving to inner

chmeliik commented on tests/test_config.py:
```diff
@@ -0,0 +1,1470 @@
+        (False, None),
+        (False, 'organization'),
+    ])
+    def test_set_source_registry(self, tmpdir, dockerfile_images, organization):
+        config = REQUIRED_CONFIG
+
+        if organization:
+            config += "\nregistries_organization: " + organization
+
+        filename = os.path.join(str(tmpdir), 'config.yaml')
+        with open(filename, 'w') as fp:
+            fp.write(dedent(config))
+        workflow = self.prepare()
```
To me this looks like the only test that actually needs a `workflow` object, could it be removed from the other tests?

rcerven replied:
removed

chmeliik reviewed (commented):
config.py LGTM, minor comments for unit tests

I only skimmed through the rest of the changes, but they did look OK :smile:

chmeliik commented:
Note on the failing coverage check: I believe most (if not all) the code that lost coverage will be removed anyway

rcerven commented:
osbs-docs https://github.com/containerbuildsystem/osbs-docs/pull/194

rcerven commented:
rn updated
